torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -1389,13 +1403,96 @@ void ONNXShapeTypeInference(
 
 } // namespace
 
+std::pair<bool, bool> IsShapeReliable(Node* n) {
+  auto reliable = true;
+  auto complete = true;
+  for (auto input : n->inputs()) {
+    if (ConstantValueMap::HasTypeReliable(input->debugName())) {
+      auto reliable_value =
+          ConstantValueMap::GetTypeReliable(input->debugName()).value();
+      reliable = reliable && reliable_value;
+    } else {
+      reliable = false;
+    }
+    if (auto pt = input->type()->cast<TensorType>()) {
+      if (!pt->sizes().isComplete()) {
+        complete = false;
+      }
+    }
+  }
+  return std::make_pair(reliable, complete);
+}
+
+// There is no need to put onnx type here, but we need this
+// for some legacy tests when onnx_shape_inference=False.
+static std::unordered_set<std::string> nodeTypeReliableForTracer = {
+    "prim::ListConstruct",
+    "onnx::Cast",
+    "onnx::Constant",
+    "com.microsoft::Gelu"};
+
+void UpdateReliable(
+    torch::jit::Value* output,
+    const std::pair<bool, bool>& inferred_type_reliable) {
+  auto inferred = false;
+  if (ConstantValueMap::HasUseInferredType(output->debugName())) {
+    inferred =
+        ConstantValueMap::GetUseInferredType(output->debugName()).value();
+  }
+  auto isTypeReliableForTracer =
+      nodeTypeReliableForTracer.find(
+          output->node()->kind().toDisplayString()) !=
+      nodeTypeReliableForTracer.end();
+  if (!inferred && !isTypeReliableForTracer &&
+      !output->node()->kind().is_onnx()) {
+    std::cerr
+        << "WARNING: The shape inference of "
+        << output->node()->kind().toDisplayString()
+        << " type is missing, so it may result in wrong shape inference for the exported graph. "
+        << "Please consider adding it in symbolic function." << std::endl;
```

Comment:
Would be awesome to update onnx doc with this PR, and link to here ... :)

Reply:
Added several sentences in onnx doc in this PR
```python
Please also consider adding shape inference logic when you regiester a custom symbolic function 
via setType API. This can help the exporter to obtain correct shape inference.
An example of setType is test_aten_embedding_2 in test_operators.py.
Although it is not required to add shape inference logic,
the exporter emits a warning message if it is not added.
```